src/GoTrueClient.ts

```diff
@@ -1056,148 +1062,121 @@ export default class GoTrueClient {
-          xform: _sessionResponse,
-          headers: this.headers,
-        }
-      )
-      if (error || !data) return { data: { user: null, session: null }, error }
-      if (data.session) {
-        await this._saveSession(data.session)
-        this._notifyAllSubscribers('MFA_CHALLENGE_VERIFIED', data.session)
-      }
-      return { data, error }
-    } catch (error) {
-      if (isAuthError(error)) {
-        return { data: { user: null, session: null }, error }
-      }
+  private async _verify(params: MFAVerifyParams): Promise<AuthMFAVerifyResponse> {
+    const { data: sessionData, error: sessionError } = await this.getSession()
+    if (sessionError) {
+      return { data: null, error: sessionError }
+    }
 
-      throw error
+    const { data, error } = await _request(
+      this.fetch,
+      'POST',
+      `${this.url}/factors/${params.factorId}/verify`,
+      {
+        body: { code: params.code, challenge_id: params.challengeId },
+        headers: this.headers,
+        jwt: sessionData?.session?.access_token,
+      }
+    )
+    if (error) {
+      return { data: null, error }
     }
+
+    await this._saveSession({
+      expires_at: Math.round(Date.now() / 1000) + data.expires_in,
+      ...data,
+    })
+    this._notifyAllSubscribers('MFA_CHALLENGE_VERIFIED', data)
+
+    return { data, error }
   }
+
   /**
    * Creates a challenge which a user can verify against
    * @param factorID System assigned identifier for authenticator device as returned by enroll
    */
-  private async _challenge(params: MFAChallengeParams) {
-    const { data, error } = await this.getUser()
-    if (error) throw error
-    const user: User = data.user
-
-    try {
-      return await _request(
-        this.fetch,
-        'POST',
-        `${this.url}/user/${user.id}/factor/${params.factorId}/challenge`,
-        {
-          body: {},
-          headers: this.headers,
-        }
-      )
-    } catch (error) {
-      if (isAuthError(error)) {
-        return { data: null, error }
-      }
-
-      throw error
+  private async _challenge(params: MFAChallengeParams): Promise<AuthMFAChallengeResponse> {
+    const { data: sessionData, error: sessionError } = await this.getSession()
+    if (sessionError) {
+      return { data: null, error: sessionError }
     }
+
+    return await _request(this.fetch, 'POST', `${this.url}/factors/${params.factorId}/challenge`, {
+      headers: this.headers,
+      jwt: sessionData?.session?.access_token,
+    })
   }
 
   /**
    * Displays all devices for a given user
    */
-  private async _listFactors() {
-    const { data, error } = await this.getUser()
-    if (error) throw error
-    return { data: { factors: data.user?.factors ?? [] }, error: error }
+  private async _listFactors(): Promise<AuthMFAListFactorsResponse> {
+    const { data: sessionData, error: sessionError } = await this.getSession()
+    if (sessionError) {
+      return { data: null, error: sessionError }
```

Comment:
```suggestion
      return { data: { all: [], totp: [] }, error: sessionError }
```

Reply:
slight improvement in DX here because data will always be an object with those keys instead of object | null